lib/ansible/module_utils/fortios.py

```diff
@@ -198,3 +208,819 @@ def get_error_infos(self, cli_errors):
+    default_ignore_params=dict(type='list'),
+    endpoint_information=dict(type='dict')
+)
+
+
+def connection_handler(func):
+    def func_wrapper(*args, **kwargs):
+        api = args[0]
+        if func.__name__ != "_login":
+            try:
+                return func(*args, **kwargs)
+            except ConnectionError:
+                try:
+                    api.fail("Connection to API endpoint %s failed with status %s" % (
+                        api._endpoint, api.response.status_code))
+                except AttributeError:
+                    api.fail(
+                        "Connection to API endpoint %s failed with no response." % api._endpoint)
+        else:
+            try:
+                return func(*args, **kwargs)
+            except ConnectionError as e:
+                try:
+                    api.fail("Failed to login to endpoint %s with status %s" % (
+                        api._endpoint, api.response.status_code))
+                except AttributeError:
+                    api._module = AnsibleModule(
+                        {}, bypass_checks=True, check_invalid_arguments=False)
+                    api.fail("Failed to login to endpoint: %s" % e)
+
+    return func_wrapper
+
+
+def return_handler(func):
+    def return_wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except (ValueError, TypeError) as e:
+            api = args[0]
+            failure_msg = "Empty response received from endpoint %s for API.%s"
+            path = args[1]
+            api.fail(failure_msg % (path, func.__name__))
+
+    return return_wrapper
+
+
+class API(object):
+
+    def __init__(self, api_info):
+
+        if not HAS_REQUESTS:
+            self.fail(
+                'Could not import the python library requests required by this module.')
+
+        self._fortigate_original_config = None
+        self._fortigate_current_config = None
+
+        self._params = _load_params()
+        self._vdom = self._params.get('vdom', "root")
+
+        self._arg_spec_filename = "FortiosAPIArgSpecs.json"
+
+        self._api_info = api_info
+        if api_info.get('from_configs') is True:
+            self._api_info = self._params['endpoint_information']
+
+        self._match_ignore_params = self._api_info.get(
+            'match_ignore_params', [])
+        if self._match_ignore_params:
+            self._match_ignore_params.append('name')
+
+        if isinstance(self._api_info["endpoint"], list):
+            self._endpoint = '/'.join(self._api_info["endpoint"])
+        else:
+            self._endpoint = self._api_info["endpoint"]
+
+        if self._params['conn_params'].get('disable_warnings', False):
+            requests.packages.urllib3.disable_warnings()
+        self._verify = self._params['conn_params'].get('verify', True)
+        self._secure = self._params['conn_params'].get('secure', True)
+        self.proxies = self._params['conn_params'].get('proxies')
```

Comment:
I don't believe this parameter should exist.  I believe we should rely on the ability for libraries to use environment variables for `http_proxy` and `https_proxy`.  Both `ansible.module_utils.urls` and `requests` can both utilize these environment vars.

Setting these values via the `environment` keyword on a task is accepted.  The module should not have a deviating method for applying proxies.

Reply:
Just want to verify - will it be possible to apply these environment variables per host in the inventory file? The idea is that you may have multiple firewalls, some of which need proxy (or even different proxies) to be managed and others don't.

Reply:
This should be modeled in the same way that all other modules within ansible currently work.  Variables can be set per host and applied to tasks.